feat(core): use faster hash function in hash tables on ILP hot path

puzpuzpuz opened the pull request:
This is an experimental PR, so any suggestions and comments are welcome.

* Replaces `Chars#hashCode()` with `Hash#hashMem()` in `DirectByteCharSequence*HashMap` maps.
* Replaces xxHash64 in `Hash#hashMem()` with a simple polynomial hash function based on the hash function from this [blog post](https://vanilla-java.github.io/2018/08/15/Looking-at-randomness-and-performance-for-hash-codes.html) by Peter Lawrey. The new function has lower quality than xxHash64, but it's much faster and much more compact which is extremely important for a hash function to be used in in-memory hash tables. This change affects `FastMap` and `CompactMap`.
* Migrates `SymbolCache` to `DirectByteCharSequenceIntHashMap`.
* Introduces `ByteCharSequence` which is an on-heap version of `DirectByteCharSequence`. This allowed getting rid of ugly UTF8 re-encoded `String`s in the code base, so both `ByteCharSequence*HashMaps` are now much less error-prone. Another positive effect is that char sequence comparison in the maps got faster. That's because `Chars#equals(DirectByteCharSequence, ByteCharSequence)` and `Chars#equals(ByteCharSequence, ByteCharSequence)` compare 4 bytes per iteration instead of a single one.
* Removes `Unsafe#swapEndianess()` in favor of `Long#reverseBytes()`.
* Increases timeouts in `IODispatcherTest#test*QueryTimeoutResetOnEachQuery()` tests.

Hash code benchmark against Java's standard hash function for `j.l.String`:
```
Benchmark                                                             (len)  Mode  Cnt   Score    Error  Units
CharSequenceHashFunctionBenchmark.testHashMemDirectByteCharSequence      7  avgt    3  2.929 ± 0.175  ns/op
CharSequenceHashFunctionBenchmark.testHashMemDirectByteCharSequence     15  avgt    3  3.684 ± 0.100  ns/op
CharSequenceHashFunctionBenchmark.testHashMemDirectByteCharSequence     31  avgt    3  5.567 ± 0.399  ns/op
CharSequenceHashFunctionBenchmark.testHashMemDirectByteCharSequence     63  avgt    3  8.930 ± 0.172  ns/op
CharSequenceHashFunctionBenchmark.testStandardDirectByteCharSequence      7  avgt    3   5.805 ±  0.534  ns/op
CharSequenceHashFunctionBenchmark.testStandardDirectByteCharSequence     15  avgt    3  11.239 ±  9.519  ns/op
CharSequenceHashFunctionBenchmark.testStandardDirectByteCharSequence     31  avgt    3  23.454 ±  0.897  ns/op
CharSequenceHashFunctionBenchmark.testStandardDirectByteCharSequence     63  avgt    3  52.815 ±  1.304  ns/op
```

Benchmark for `DirectByteCharSequenceIntHashMap` (`Hash#hashMem()`) vs. `CharSequenceIntHashMap` (standard hash function):
```
Benchmark                                             (len)  (n)  Mode  Cnt      Score     Error  Units
ByteCharSequenceIntHashMapBenchmark.testDirectMap         7   16  avgt    3     87.208 ±  11.173  ns/op
ByteCharSequenceIntHashMapBenchmark.testDirectMap         7  256  avgt    3   1539.001 ±  28.377  ns/op
ByteCharSequenceIntHashMapBenchmark.testDirectMap        15   16  avgt    3    126.760 ±   1.158  ns/op
ByteCharSequenceIntHashMapBenchmark.testDirectMap        15  256  avgt    3   1859.707 ±  12.316  ns/op
ByteCharSequenceIntHashMapBenchmark.testDirectMap        31   16  avgt    3    151.213 ±   3.643  ns/op
ByteCharSequenceIntHashMapBenchmark.testDirectMap        31  256  avgt    3   2344.601 ±  46.161  ns/op
ByteCharSequenceIntHashMapBenchmark.testDirectMap        63   16  avgt    3    219.332 ±  27.181  ns/op
ByteCharSequenceIntHashMapBenchmark.testDirectMap        63  256  avgt    3   3244.319 ±  49.826  ns/op
ByteCharSequenceIntHashMapBenchmark.testNonDirectMap      7   16  avgt    3    195.619 ±   5.376  ns/op
ByteCharSequenceIntHashMapBenchmark.testNonDirectMap      7  256  avgt    3   2853.950 ±  67.139  ns/op
ByteCharSequenceIntHashMapBenchmark.testNonDirectMap     15   16  avgt    3    329.766 ±  14.430  ns/op
ByteCharSequenceIntHashMapBenchmark.testNonDirectMap     15  256  avgt    3   4028.574 ± 220.291  ns/op
ByteCharSequenceIntHashMapBenchmark.testNonDirectMap     31   16  avgt    3    447.999 ±   3.538  ns/op
ByteCharSequenceIntHashMapBenchmark.testNonDirectMap     31  256  avgt    3   6979.639 ±  56.590  ns/op
ByteCharSequenceIntHashMapBenchmark.testNonDirectMap     63   16  avgt    3    850.159 ±  25.992  ns/op
ByteCharSequenceIntHashMapBenchmark.testNonDirectMap     63  256  avgt    3  14620.540 ±  85.237  ns/op
```

ideoma commented:
#### [PR Coverage check]

:disappointed: **fail** : 0 / 1 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/str/DirectByteCharSequence.java|0|1|00.00%|

bluestreak01 commented:
FYI, the current hash function is compatible with that generated by `String` class. With function being different `String` and our `CharSequence` impls will behave different in maps other than those dedicated exclusively to `DirectByteCharSequence`

puzpuzpuz commented:
Yes, I'm aware of the origin of that function. Yet, I don't see any issues with the hot path maps using a faster hash function. Do you see any?

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 41 / 41 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Hash.java|20|20|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceIntHashMap.java|21|21|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 80 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Hash.java|18|18|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceObjHashMap.java|33|33|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceIntHashMap.java|29|29|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 80 / 80 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Hash.java|18|18|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceObjHashMap.java|33|33|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceIntHashMap.java|29|29|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 89 / 89 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Hash.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceObjHashMap.java|33|33|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceIntHashMap.java|29|29|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

ideoma commented on core/src/main/java/io/questdb/std/Hash.java:
```diff
@@ -98,6 +101,14 @@ public static int spread(int h) {
+     * are also supported, yet with higher chances of hash code collisions.
+     */
+    public static long xxHash64(String str) {
+        return xxHash64(0, str.length(), 0, stringUtf8MemoryAccessor.get().of(str));
```
Thread locals are usually quite expensive.
Are you sure you benched this code path and there is still a benefit?

puzpuzpuz replied:
Yup, it's benched. `DirectByteCharSequence*HashMap` now cache stored hash codes, so string hashe codes are only calculated once. On the hot path we only do look-ups for `DirectByteCharSequence`s, so that's what we optimize for.

ideoma replied:
Please move `DirectByteCharSequenceHashMap` into ILP package and make it package private, seems that it cannot be used generically and `std` is not a good place for it

ideoma replied:
Also since it's not for the hot path I would move ThreadLocal out of `Hash` class into `DirectByteCharSequenceHashMap`

ideoma replied:
Why you did not want to have a version with inlined accessor for Strings and regular CharSequence? It could be useful in other hash maps since it's faster.

puzpuzpuz replied:
> Why you did not want to have a version with inlined accessor for Strings and regular CharSequence? It could be useful in other hash maps since it's faster.

I didn't want to end up with N versions of the same complex hash function. JIT compiler should be able to inline the accessor.

puzpuzpuz replied:
Moved the maps to `.cutlass.line.tcp` in 7bc20de53dfc67ead372529b6de385479e083e01 as well as the thread-local and `StringUtf8MemoryAccessor`. Had to keep `DirectByteCharSequenceIntHashMap` public since it's used in the JMH benchmark.

puzpuzpuz replied:
One more follow-up: removed the thread local completely in c36de175813fd349c063c77d82fa2a7f3897c2c9

ideoma commented on benchmarks/src/main/java/org/questdb/CharSequenceHashFunctionBenchmark.java:
```diff
@@ -0,0 +1,83 @@
+@State(Scope.Benchmark)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+public class CharSequenceHashFunctionBenchmark {
```
Can you bench Strings too please?

puzpuzpuz replied:
`String` hash code calculation is not on the hot path. See https://github.com/questdb/questdb/pull/2868#discussion_r1054773042. I can include it here, but it wouldn't be very illustrative.

puzpuzpuz replied:
Added the `String` benchmark in 29a795d5361d828080b7a8c2626c6de8a861fce3. The results are in the PR description (see `testXXHashString`)

ideoma commented on core/src/test/java/io/questdb/std/DirectByteCharSequenceObjHashMapTest.java:
```diff
@@ -138,9 +141,34 @@ public void testHashMapCompatibility() {
+            final StringSink sink = new StringSink();
+            final String utf8Str = sink.put(dbcs).toString();
+            map.put(utf8Str, 42);
+            Assert.assertEquals(42, (int) map.get(dbcs));
```
Can you test other methods too, so that putting by CharSequence and getting by string is also same as well as remove etc?

puzpuzpuz replied:
Putting by `CharSequence` is not supported by `DirectByteCharSequence*HashMap` collections. Added extra checks and a remove call in 29a795d5361d828080b7a8c2626c6de8a861fce3

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 89 / 89 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Hash.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceObjHashMap.java|33|33|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceIntHashMap.java|29|29|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

ideoma commented on core/src/test/java/io/questdb/std/DirectByteCharSequenceIntHashMapTest.java:
```diff
@@ -138,9 +141,39 @@ public void testHashMapCompatibility() {
+            Assert.assertEquals(42, map.get(dbcs));
+            Assert.assertEquals(42, map.get(utf8Str));
 
+            Assert.assertEquals(1, map.size());
```
Can you also put by `dbcs` and get by `utf8Str`? Can you find and test hash collisions as well?

puzpuzpuz replied:
> put by `dbcs`

This is not supported by these maps. We only put strings into them in the production code, so adding such overload would be synthetic and test-only.

puzpuzpuz replied:
> Can you find and test hash collisions as well?

Not sure if I follow this question. Hash collisions should be analyzed against the hash function, not the end hash table. [This repo](https://github.com/rurban/smhasher) has results for xxHash64 which is the hash function we use. Am I missing something?

puzpuzpuz replied:
> Perhaps it's indirectly tested in ILP tests?

Yes, that's the case, but adding more cases here is a good idea. Let me do that.

puzpuzpuz replied:
Done in efbac9600c69e97078e6c73e7407d3515c9deeb3

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 92 (98.91%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceIntHashMap.java|53|54|98.15%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceObjHashMap.java|34|34|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 91 / 92 (98.91%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceIntHashMap.java|53|54|98.15%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceObjHashMap.java|34|34|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

puzpuzpuz commented:
@ideoma I've migrated `SymbolCache` to `DirectByteCharSequenceIntHashMap` in 8c1a733a873d889a83811136b8d0c7e68b7fb630 and 6cbc8d003a39d8d6fd476fd554700a49c3bb8748

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 114 / 115 (99.13%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceIntHashMap.java|66|67|98.51%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceObjHashMap.java|34|34|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

puzpuzpuz commented on core/src/main/java/io/questdb/cutlass/line/tcp/DirectByteCharSequenceIntHashMap.java:
```diff
@@ -251,19 +298,78 @@ private void putAt0(int index, String key, int value) {
+        }
+
+        public int length() {
+            return str.length();
```
Used since c36de175813fd349c063c77d82fa2a7f3897c2c9

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 116 / 116 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceObjHashMap.java|35|35|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceIntHashMap.java|67|67|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 116 / 116 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceObjHashMap.java|35|35|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceIntHashMap.java|67|67|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 116 / 116 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceObjHashMap.java|35|35|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceIntHashMap.java|67|67|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

puzpuzpuz commented:
@ideoma I've noticed that xxHash64 doesn't shine when it comes to short strings. That's because it's rather big code-wise, so C2 doesn't inline it and, even if it would inline it, it wouldn't be very i-cache friendly. So, for shorter strings it's on par with the standard hash function:

```
Benchmark                                                   (len)  (n)  Mode  Cnt     Score     Error  Units
DirectByteCharSequenceIntHashMapBenchmark.testDirectMap        10   16  avgt    3   251.350 ±  48.081  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testDirectMap        10   64  avgt    3   859.347 ±  47.275  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testDirectMap        15   16  avgt    3   302.386 ±  11.256  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testDirectMap        15   64  avgt    3  1084.782 ±  37.464  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testDirectMap        20   16  avgt    3   333.613 ±  14.726  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testDirectMap        20   64  avgt    3   989.238 ± 301.036  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testNonDirectMap     10   16  avgt    3   234.063 ±   3.634  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testNonDirectMap     10   64  avgt    3   711.534 ±  63.199  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testNonDirectMap     15   16  avgt    3   384.577 ±  26.749  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testNonDirectMap     15   64  avgt    3   938.704 ±   3.607  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testNonDirectMap     20   16  avgt    3   366.003 ±   1.735  ns/op
DirectByteCharSequenceIntHashMapBenchmark.testNonDirectMap     20   64  avgt    3  1159.596 ± 106.407  ns/op
```

I'm marking the PR as a draft to update it later. I have a few ideas I'd like to try this evening.

puzpuzpuz commented:
> I'm marking the PR as a draft to update it later. I have a few ideas I'd like to try this evening.

The PR is ready for review once again. The main changes are in 728205f5696e08629d518893b696c9ba784c5a95. I got rid of xxHash64 and described the motivation in the description. The benchmark results are also up to date.

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 133 / 133 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceObjHashMap.java|34|34|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|41|41|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/DirectByteCharSequenceIntHashMap.java|42|42|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMap.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|

puzpuzpuz commented:
I introduced `ByteCharSequence` which is an on-heap version of `DirectByteCharSequence` in a5257c48cf668aa4999fab45e74d5b07b08c21cb

This allowed getting rid of ugly UTF8 re-encoded `String`s in the code base, so both `ByteCharSequence*HashMap`s are now much less error-prone. That's why I moved them back to `io.questdb.std` package. Another positive side effect is saved on-heap memory, especially when compared with non-ASCII strings.

Finally, both `ByteCharSequence*HashMap` got 20-50% faster due to more efficient comparison. `Chars#equals(DirectByteCharSequence, ByteCharSequence)` and `Chars#equals(ByteCharSequence, ByteCharSequence)` compare 4 bytes per iteration instead of a single one. I've updated the benchmark results in the PR description.

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 188 / 238 (78.99%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Chars.java|20|68|29.41%|
|:large_blue_circle:|io/questdb/std/str/ByteCharSequence.java|11|13|84.62%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceIntHashMap.java|46|46|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceObjHashMap.java|40|40|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|35|35|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpNetworkIOJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMap.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|6|6|100.00%|

puzpuzpuz commented:
Added hash code caching to FastMap to avoid high latency due to redundant LOADs and key comparisons on look-up in eca30bac5b695859a95862cc00dea384009549bb

FastMap benchmarks on `master`:
```
Benchmark                         Mode  Cnt  Score   Error  Units
QMapWriteBenchmark.baseline       avgt    3  0.170 ± 0.005  us/op
QMapWriteBenchmark.testDirectMap  avgt    3  0.414 ± 0.005  us/op
QMapWriteBenchmark.testHashMap    avgt    3  0.581 ± 0.093  us/op
QMapWriteBenchmark.testQMap       avgt    3  0.427 ± 0.009  us/op

Benchmark                             Mode  Cnt  Score    Error  Units
QMapWriteLongBenchmark.baseline       avgt    5  0.002 ±  0.001  us/op
QMapWriteLongBenchmark.testDirectMap  avgt    5  0.130 ±  0.002  us/op
QMapWriteLongBenchmark.testQMap       avgt    5  0.169 ±  0.001  us/op

Benchmark                                 Mode  Cnt  Score   Error  Units
QMapReadRandomKeyBenchmark.baseline       avgt    3  0.199 ± 0.008  us/op
QMapReadRandomKeyBenchmark.testDirectMap  avgt    3  0.418 ± 0.003  us/op
QMapReadRandomKeyBenchmark.testHashMap    avgt    3  0.400 ± 0.019  us/op
QMapReadRandomKeyBenchmark.testQMap       avgt    3  0.415 ± 0.014  us/op

Benchmark                                     Mode  Cnt  Score    Error  Units
QMapReadSequentialKeyBenchmark.baseline       avgt    3  0.005 ±  0.001  us/op
QMapReadSequentialKeyBenchmark.testDirectMap  avgt    3  0.042 ±  0.001  us/op
QMapReadSequentialKeyBenchmark.testHashMap    avgt    3  0.021 ±  0.001  us/op
QMapReadSequentialKeyBenchmark.testQMap       avgt    3  0.046 ±  0.019  us/op
```

FastMap benchmarks on this branch:
```
Benchmark                         Mode  Cnt  Score   Error  Units
QMapWriteBenchmark.baseline       avgt    3  0.168 ± 0.002  us/op
QMapWriteBenchmark.testDirectMap  avgt    3  0.360 ± 0.006  us/op
QMapWriteBenchmark.testHashMap    avgt    3  0.577 ± 0.073  us/op
QMapWriteBenchmark.testQMap       avgt    3  0.428 ± 0.004  us/op

Benchmark                             Mode  Cnt  Score    Error  Units
QMapWriteLongBenchmark.baseline       avgt    5  0.002 ±  0.001  us/op
QMapWriteLongBenchmark.testDirectMap  avgt    5  0.120 ±  0.002  us/op
QMapWriteLongBenchmark.testQMap       avgt    5  0.163 ±  0.006  us/op

Benchmark                                 Mode  Cnt  Score   Error  Units
QMapReadRandomKeyBenchmark.baseline       avgt    3  0.190 ± 0.011  us/op
QMapReadRandomKeyBenchmark.testDirectMap  avgt    3  0.413 ± 0.003  us/op
QMapReadRandomKeyBenchmark.testHashMap    avgt    3  0.395 ± 0.003  us/op
QMapReadRandomKeyBenchmark.testQMap       avgt    3  0.411 ± 0.024  us/op

Benchmark                                     Mode  Cnt  Score    Error  Units
QMapReadSequentialKeyBenchmark.baseline       avgt    3  0.005 ±  0.001  us/op
QMapReadSequentialKeyBenchmark.testDirectMap  avgt    3  0.037 ±  0.005  us/op
QMapReadSequentialKeyBenchmark.testHashMap    avgt    3  0.021 ±  0.004  us/op
QMapReadSequentialKeyBenchmark.testQMap       avgt    3  0.040 ±  0.001  us/op
```

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 258 / 308 (83.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Chars.java|20|68|29.41%|
|:large_blue_circle:|io/questdb/std/str/ByteCharSequence.java|11|13|84.62%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceIntHashMap.java|46|46|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceObjHashMap.java|40|40|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|35|35|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpNetworkIOJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|72|72|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMap.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|6|6|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 271 / 319 (84.95%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Chars.java|22|68|32.35%|
|:large_blue_circle:|io/questdb/std/str/ByteCharSequence.java|12|14|85.71%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceIntHashMap.java|46|46|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceObjHashMap.java|40|40|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|43|43|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpNetworkIOJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|72|72|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMap.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|6|6|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 271 / 319 (84.95%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Chars.java|22|68|32.35%|
|:large_blue_circle:|io/questdb/std/str/ByteCharSequence.java|12|14|85.71%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceIntHashMap.java|46|46|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceObjHashMap.java|40|40|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|43|43|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpNetworkIOJob.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/FastMap.java|72|72|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMap.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|6|6|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 205 / 254 (80.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Chars.java|22|68|32.35%|
|:large_blue_circle:|io/questdb/std/str/ByteCharSequence.java|12|14|85.71%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceIntHashMap.java|45|46|97.83%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceObjHashMap.java|40|40|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|25|25|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpNetworkIOJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMap.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|6|6|100.00%|

ideoma commented:
#### [PR Coverage check]

:heart_eyes: **pass** : 206 / 254 (81.10%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/Chars.java|23|68|33.82%|
|:large_blue_circle:|io/questdb/std/str/ByteCharSequence.java|12|14|85.71%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceIntHashMap.java|45|46|97.83%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpEventBuffer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementScheduler.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/wal/WalWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|27|27|100.00%|
|:large_blue_circle:|io/questdb/std/ByteCharSequenceObjHashMap.java|40|40|100.00%|
|:large_blue_circle:|io/questdb/std/Hash.java|25|25|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpNetworkIOJob.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Unsafe.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/CompactMap.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/fastdouble/FastDoubleSwar.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/LineTcpMeasurementEvent.java|6|6|100.00%|
